datafusion/common/src/config.rs

```diff
@@ -338,6 +338,19 @@ config_namespace! {
         /// if the source of statistics is accurate.
         /// We plan to make this the default in the future.
         pub use_row_number_estimates_to_optimize_partitioning: bool, default = false
+
+        /// Should DataFusion use the the blocked approach to manage the groups
+        /// values and their related states in accumulators. By default, the single
+        /// approach will be used, values are managed within a single large block
+        /// (can think of it as a Vec). As this block grows, it often triggers
+        /// numerous copies, resulting in poor performance.
+        /// If setting this flag to `true`, the blocked approach will be used.
+        /// And the blocked approach allocates capacity for the block
+        /// based on a predefined block size firstly. When the block reaches its limit,
+        /// we allocate a new block (also with the same predefined block size based capacity)
+        // instead of expanding the current one and copying the data.
+        /// We plan to make this the default in the future when tests are enough.
+        pub enable_aggregation_intermediate_states_blocked_approach: bool, default = false
```

Comment:
Making it default / not a config should reduce complexity...

Reply:
I am just not sure if we should enable it by default at the beginning.
But as I see, current tests may be enough to keep the correntness of this optimization?

Reply:
I think coverage is high enough, so if we can show that the test pass, and benchmarks don't show (large) regressions, there is no real downside for enabling it.
Maybe we can add some extra tests for e.g. testing the reduced memory usage with this approach.

Reply:
This pr is too stale to continue developing, I am pushing it forward in #15591

Reply:
Perfect! Let's close this one